File: threading/cpp/09_barrier_latch.cpp
/*
 * QUESTION 9: Barrier and Latch (Databricks/Rubrik)
 * ==================================================
 * 
 * Problem: Implement synchronization primitives for phase-based computation.
 * 
 * Key Concepts: std::barrier (C++20), std::latch (C++20), custom implementations
 * 
 * Compile: g++ -std=c++20 -pthread 09_barrier_latch.cpp -o barrier_latch
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include <chrono>
#include <barrier>
#include <latch>

// Custom Barrier (for C++11/14/17)
class Barrier {
    /*
     * EXPLANATION:
     * All N threads must arrive before any can proceed.
     * 
     * Use Cases:
     * - Parallel algorithms with phases
     * - MapReduce: All mappers finish before reduce
     */
private:
    std::mutex mutex_;
    std::condition_variable cv_;
    size_t parties_;
    size_t count_;
    size_t generation_ = 0;

public:
    explicit Barrier(size_t parties) : parties_(parties), count_(parties) {}

    void wait() {
        std::unique_lock<std::mutex> lock(mutex_);
        size_t gen = generation_;
        
        if (--count_ == 0) {
            // Last to arrive - reset and wake all
            count_ = parties_;
            ++generation_;
            cv_.notify_all();
        } else {
            // Wait for others
            cv_.wait(lock, [this, gen] { return gen != generation_; });
        }
    }
};

// Custom CountDownLatch (for C++11/14/17)
class CountDownLatch {
    /*
     * EXPLANATION:
     * - Initialize with count N
     * - Some threads await(), others count_down()
     * - When count=0, all waiters proceed
     * - NOT reusable (one-shot)
     */
private:
    std::mutex mutex_;
    std::condition_variable cv_;
    size_t count_;

public:
    explicit CountDownLatch(size_t count) : count_(count) {}

    void count_down() {
        std::lock_guard<std::mutex> lock(mutex_);
        if (count_ > 0 && --count_ == 0) {
            cv_.notify_all();
        }
    }

    void wait() {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return count_ == 0; });
    }
};

void barrier_demo() {
    std::cout << "=== Barrier Demo ===" << std::endl;
    
    constexpr int NUM_WORKERS = 4;
    constexpr int NUM_PHASES = 3;
    Barrier barrier(NUM_WORKERS);
    std::mutex cout_mutex;
    
    auto worker = [&](int id) {
        for (int phase = 0; phase < NUM_PHASES; ++phase) {
            {
                std::lock_guard<std::mutex> lock(cout_mutex);
                std::cout << "Worker " << id << " phase " << phase << std::endl;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100 * (id + 1)));
            barrier.wait();
        }
    };
    
    std::vector<std::thread> threads;
    for (int i = 0; i < NUM_WORKERS; ++i) {
        threads.emplace_back(worker, i);
    }
    for (auto& t : threads) t.join();
}

void latch_demo() {
    std::cout << "\n=== CountDownLatch Demo ===" << std::endl;
    
    CountDownLatch latch(3);
    
    auto init_service = [&](const std::string& name, int delay_ms) {
        std::cout << name << " initializing..." << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms));
        std::cout << name << " ready!" << std::endl;
        latch.count_down();
    };
    
    std::thread t1(init_service, "Database", 300);
    std::thread t2(init_service, "Cache", 200);
    std::thread t3(init_service, "MessageQueue", 400);
    
    std::cout << "Main: Waiting for services..." << std::endl;
    latch.wait();
    std::cout << "Main: All services ready!" << std::endl;
    
    t1.join(); t2.join(); t3.join();
}

// C++20 std::barrier and std::latch demo
void cpp20_demo() {
    std::cout << "\n=== C++20 std::barrier ===" << std::endl;
    
    auto on_completion = []() noexcept {
        std::cout << "Phase complete!" << std::endl;
    };
    
    std::barrier sync_point(3, on_completion);
    
    auto task = [&](int id) {
        std::cout << "Thread " << id << " working" << std::endl;
        sync_point.arrive_and_wait();
    };
    
    std::thread t1(task, 1), t2(task, 2), t3(task, 3);
    t1.join(); t2.join(); t3.join();
}

int main() {
    barrier_demo();
    latch_demo();
    cpp20_demo();
    return 0;
}

File: threading/python/09_barrier_latch.py
"""
QUESTION 9: Barrier and CountDownLatch (Databricks/Rubrik)
==========================================================

Problem: Implement synchronization primitives for phase-based computation.

Key Concepts: Synchronization points, MapReduce patterns
"""

import threading
import time


class Barrier:
    """
    Reusable Barrier Implementation.
    
    EXPLANATION:
    All N threads must arrive before any can proceed.
    
    Use Cases:
    - Parallel algorithms with phases
    - MapReduce: All mappers finish before reduce
    - Simulations: All update before render
    """
    
    def __init__(self, parties: int):
        self.parties = parties
        self.count = parties
        self.generation = 0
        self._condition = threading.Condition()
    
    def wait(self) -> int:
        """Wait for all parties. Returns arrival index."""
        with self._condition:
            gen = self.generation
            index = self.parties - self.count
            self.count -= 1
            
            if self.count == 0:
                # Last to arrive - reset and wake all
                self.count = self.parties
                self.generation += 1
                self._condition.notify_all()
            else:
                while gen == self.generation:
                    self._condition.wait()
            
            return index


class CountDownLatch:
    """
    One-shot CountDownLatch.
    
    EXPLANATION:
    - Initialize with count N
    - Some threads await(), others count_down()
    - When count=0, all waiters proceed
    - NOT reusable
    
    DIFFERENCE from Barrier:
    - Barrier: All threads wait AND signal
    - Latch: Some wait, others signal
    """
    
    def __init__(self, count: int):
        self.count = count
        self._condition = threading.Condition()
    
    def count_down(self):
        with self._condition:
            if self.count > 0:
                self.count -= 1
                if self.count == 0:
                    self._condition.notify_all()
    
    def await_latch(self, timeout: float = None) -> bool:
        with self._condition:
            return self._condition.wait_for(
                lambda: self.count == 0, timeout
            )


def barrier_demo():
    """Parallel matrix processing simulation."""
    NUM_WORKERS, NUM_PHASES = 4, 3
    barrier = Barrier(NUM_WORKERS)
    
    def worker(wid):
        for phase in range(NUM_PHASES):
            print(f"Worker {wid} phase {phase}")
            time.sleep(0.1 * (wid + 1))
            barrier.wait()
        print(f"Worker {wid} done")
    
    threads = [threading.Thread(target=worker, args=(i,)) for i in range(NUM_WORKERS)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()


def latch_demo():
    """Service initialization simulation."""
    latch = CountDownLatch(3)
    
    def init_service(name, duration):
        print(f"{name} initializing...")
        time.sleep(duration)
        print(f"{name} ready!")
        latch.count_down()
    
    for name, dur in [("DB", 0.3), ("Cache", 0.2), ("MQ", 0.4)]:
        threading.Thread(target=init_service, args=(name, dur)).start()
    
    print("Main: Waiting for services...")
    latch.await_latch()
    print("Main: All services ready!")


if __name__ == "__main__":
    print("=== Barrier Demo ===")
    barrier_demo()
    print("\n=== Latch Demo ===")
    latch_demo()

